### ISSUE TITLE
Mixed dataframe - array graph optimization (column projection pushdown not working if final result is an array)

### ISSUE THREAD

**Issue by jorisvandenbossche** [dask/dask#7885]
From https://github.com/geopandas/dask-geopandas/issues/78, but I made a small standalone example to illustrate the issue (see below for the full example workflow).  
If you have a workflow where you read a DataFrame from a file (eg parquet, which supports column selection), select a subset of the columns, and then perform an operation which results in an Array and not Series/DataFrame, the "column projection" pushdown (`project_columns` method of the `DataFrameIOLayer`, used in the `optimize_dataframe_getitem` function which is called in the general graph optimization step for DataFrames) is not working. 

To the extent I understand the graph optimization and debugged this use case, the reason this doesn't work in this case is because the final collection object of the workflow is a dask Array (although all layers of the HighLevelGraph except for the last ones are still DataFrame-based), and thus the optimization step at

https://github.com/dask/dask/blob/2a0020daa5ca960d9ae7f898b181d28693367a2f/dask/base.py#L333-L338

uses the Array's optimize function (`Array.__dask_optimize__` -> `dask.array.optimization.optimize`) to optimize the full graph instead of the dask DataFrame version (and it's the dataframe version that has the logic to call `optimize_dataframe_getitem` and optimize this).

We can probably work around this in dask-geopandas (eg return a 4-element Series instead of 4-element array for the `total_bounds` attribute in question), but would such mixed-layer optimizations be in scope for dask?

----

**Full example**

Consider the following dask dataframe from a Parquet dataset:

```python
import pandas as pd
import dask.dataframe as dd

df = pd.DataFrame(np.random.randn(100, 3), columns=["a", "b", "c"])
ddf = dd.from_pandas(df, npartitions=2)
ddf.to_parquet("test_column_projection")

ddf = dd.read_parquet("test_column_projection/", engine="pyarrow")
```

and I create a reduction operation (example here is to calculate min and max of one column):

```python
def reduce_chunk(s):
    return pd.Series([s.min(), s.max()], index=["min", "max"])

def reduce_aggregate(df):
    return pd.Series([df["min"].min(), df["max"].max()], index=["min", "max"])

result = ddf["a"].reduction(reduce_chunk, aggregate=reduce_aggregate)
```

Thanks to the column projection pushdown in the graph optimization, the `getitem` layer ensures that the Parquet reader only reads column "a". To ensure this for illustration purposes, I added a print statement to the parquet reading function:

```python
from dask.dataframe.io.parquet import arrow

def log_columns_keyword(f):
    """Print the value passed for the columns to read"""
    def wrapper(*args, **kwargs):
        print(f.__name__, "columns:", args[3])
        return f(*args, **kwargs)
    return wrapper

arrow._read_table_from_path = log_columns_keyword(arrow._read_table_from_path)
```

which gives:
```
In [4]: result.compute(scheduler="synchronous")
_read_table_from_path columns: ['a', '__null_dask_index__']
_read_table_from_path columns: ['a', '__null_dask_index__']
Out[4]: 
min   -1.962085
max    2.065056
dtype: float64
```

However, if I now rewrite the reduction to return an array instead of a Series:

```python
def reduce_chunk2(s):
    return np.array([s.min(), s.max()])

def reduce_aggregate2(arr):
    return np.array([arr[0::2].min(), arr[1::2].max()])

result2 = ddf["a"].reduction(reduce_chunk2, aggregate=reduce_aggregate2)
```

we get:

```
In [6]: result2.compute(scheduler="synchronous")
_read_table_from_path columns: ['a', 'b', 'c', '__null_dask_index__']
_read_table_from_path columns: ['a', 'b', 'c', '__null_dask_index__']
Out[6]: array([-1.96208513,  2.06505574])
```

So here it is reading all columns from the Parquet file, even though I am only calculating something on `ddf["a"]`.

**Comment by mrocklin** [comment]
Mixed type optimization is certainly in scope.  I think that first we need to improve typing of all of the individual layers, and then we can change dask.optimize to optimize connected components of layers based on their type, rather than based on the type of the final colleciton (which, as you point out, is not great).  

I think that right now @rjzamora is looking into more extensive typing of dataframes as a start to this process.  

As a warning though, this has been the plan for several months, and has not risen to the top of anyone's priority list.  If anyone is interested in pushing on this I'm more than happy to advise them through it.

**Comment by rjzamora** [comment]
>I think that right now @rjzamora is looking into more extensive typing of dataframes as a start to this process.

Yes, I will soon submit a draft PR demonstrating the basic changes required to enable "multi-layer" column projection in Dask-Dataframe and outlining a general design proposal.  The proposed changes are dominated by the need for all collection-based Layers to inherit from a corresponding `<Collection>Layer` class.  For Dask-Dataframe, this will make it easier to track column dependencies between layers (note that I have started working through a simple `set_index` example to explore/test this).  That work also motivates a new `Layer` method to replace itself when one of it's dependencies has changed.

**Comment by mrocklin** [comment]
My thinking is that once that is done it should be easier to segment the
graph based on layer type, and solve the problem that Joris has in this
issue.

**Comment by jrbourbeau** [comment]
Just cross-referencing another place where this came up https://github.com/dask/dask/issues/7545
